Eq instance exercise & ghcid instructions

mankyKitty opened the pull request:
Related to #19 #20 

Remove a function that if implemented could easily cause an import
cycle.

Add an exercise for implementing the `Eq` instance for `MyBTree` and
then checking that it satisfies the laws.

Leaving the completed exercises in for the moment to check my working.

Updated the readme to include some instructions for configuring ghcid to
run tests when the code is buildable. Needs finessing as there are some
wrinkles that won't be obvious.

mankyKitty committed:
Eq instance exercise & ghcid

Remove a function that if implemented could easily cause an import
cycle.

Add an exercise for implementing the `Eq` instance for `MyBTree` and
then checking that it satisfies the laws.

Leaving the completed exercise in for the moment to check my working.

Updated the readme to include some instructions for configuring ghcid to
run tests when the code is buildable. Needs finessing as there are some
wrinkles that won't be obvious.

endgame commented on level00/PropertyTests.hs:
```diff
@@ -112,7 +113,44 @@ genTree genKV = fromList <$> Gen.list (Range.linear 0 100) genKV
+-- with the laws of that typeclass. This is important as often the
+-- laws of a typeclass cannot be expressed in the types.
+--
+-- Implement the 'Eq' instance for 'MyBTree' and then complete the
```
The Eq instance is not unique- should we talk about trees being equal if they have the same keys/vals instead of identical structure?

mankyKitty replied:
Not sure how much detail to go into ? I've updated the instructions a bit...

Also does my implementation of the property look sufficient ? Bumped it to a hundred thousand tests (lulz) and it was still okay so I'm somewhat confident.

mankyKitty committed:
Update instructions for Eq instances for MyBTree

Indicate that we want participants to implement the Eq instance that
checks for a matching set of keys and values, rather than structural
equality.

Drop the Ord constraint on the `toListWithKey` function.

endgame commented on level00/PropertyTests.hs:
```diff
@@ -112,7 +113,45 @@ genTree genKV = fromList <$> Gen.list (Range.linear 0 100) genKV
+-- with the laws of that typeclass. This is important as often the
+-- laws of a typeclass cannot be expressed in the types.
+--
+-- NOTE: There are _two_ valid 'Eq' instances we could write:
```
There's also `const $ const True`, and others.

mankyKitty replied:
Are there other _useful_ instances, you could write one that loops forever, you could leave it as undefined, you could flip the arguments then compare, you could write one that compares the first argument to itself and vice versa then return the XOR of that... 

Happy to expand if there are other *valid* and *useful* instances, but I come down hard on middling "but what if I just X!" pedantry.

endgame replied:
I don't think it's pedantry. Any function from `a -> a -> Bool` that satisfies the `Eq` laws is valid, as it partitions `a` into equivalence classes. I think comparing keys only gives you a potentially-useful instance, but I haven't thought deeply about that.

I suggest this phrasing:

> The `Eq` instance from `deriving Eq` is too granular: it will consider unequal trees with the same content but different structure, which will be surprising to our callers

tonymorris replied:
Write a test that selects the instance that is intended. i.e. all others should fail the test.

mankyKitty replied:
I've done both. :) Although I suspect the 'you've written the desired instance' test will need expanding as I only counter for structural equality. Should check for key or value only based equality? I don't think participants will be that deliberately tricksy.. ? Maybe?

endgame commented on level00/PropertyTests.hs:
```diff
@@ -112,7 +113,45 @@ genTree genKV = fromList <$> Gen.list (Range.linear 0 100) genKV
+  annotate "Symmetry: x == y = y == x"
+  (x == y) === (y == x)
+
+  annotate "Transitivity: if x == y && y == x = True then x == z = True"
```
You could probably drop the `= True` from these?

mankyKitty replied:
They're copied verbatim from the `Eq` docs, which I haven't included a link to. I will expand to include a link.

endgame commented on level00/PropertyTests.hs:
```diff
@@ -112,7 +113,45 @@ genTree genKV = fromList <$> Gen.list (Range.linear 0 100) genKV
+  when (x == y && y == z) $ (x == z) === True
+
+  annotate "Substitutivity: if x == y = True and (f :: (Eq a, Eq b) => a -> b) then f x == f y = True"
+  when (x == y) $ (null x == null y) === True
```
This is not a test for the whole property. Add a commentary?

endgame replied:
It tests for only one function. Dunno what a good general property would look like; can you generate functions from trees to `Int`?

mankyKitty replied:
Bring in `hedgehog-fn` and have at it ?

endgame replied:
dooo eeet

Might be worth having some language in there to say "if you're getting along well enough, consider jumping off to level 1 instead of doing literally every exercise"?

mankyKitty replied:
Can't use hedgehog-fn after initial attempts, it also might push the weirdo-needle too far.

That bonus wording is definitely a good idea though.

mankyKitty replied:
Made `hedgehog-fn` work, it's a bit crunchy though. Would appreciate more eyes to make it palatable.

mankyKitty committed:
Add hedgehog-fn for Eq instance tests

Added hedgehog-fn for generating the (a -> Bool) function for checking
the substitutivity property. Not sure if this pushes the weirdness
needle too far as we can't use 'Char' in the value position as there
is no Generic instance for it.

Added some more notes to push people on to the other exercises and try
to keep people focused when writing the Eq instance.

mankyKitty committed:
Add example property model.

Model for property actions was burning a hole in my brain, had to put
it somewhere.

updated some comments

mankyKitty committed:
Remove prism 'bonus' level

Add a test to try to catch when structural equality is used.

mankyKitty committed:
Add more equality checks

Added more tests to ensure people aren't using key only or value only
equality.

endgame commented on level00/ExampleModelProperties.hs:
```diff
@@ -0,0 +1,99 @@
+
+-- About that generator...
+--
+-- Our model is a recursive structure, so we leverage the functions in
```
> Our model is a recursive structure. Hedgehog provides functions to manage recursive generators, lest they produce infinite structures by accident. This also allows ...

endgame commented on level00/MyBTree.hs:
```diff
@@ -9,19 +10,26 @@ module MyBTree
 
 fromList :: (Foldable f, Ord k) => f (k,a) -> MyBTree k a
-fromList = foldr (uncurry insert) Empty
+fromList = foldl (\m (k, a) -> insert k a m) Empty
```
`foldl'`?

mankyKitty commented on level00/ExampleModelProperties.hs:
```diff
@@ -0,0 +1,99 @@
+eval EmptyModel = Empty
+
+-- We also need to be able to evaluate the actions against a
+-- reference. For this we will use the 'Set' type from the
```
Should be `Map`

mankyKitty committed:
Remove implementations.

Start process of converting these properties into exercises.

Use `foldl'` over `foldl`.

Refer to Map not Set as the reference type has changed.

mankyKitty committed:
Add missing import
